### ISSUE TITLE
[Bug]Same function declaration will cause SyntaxError

### ISSUE THREAD

**Issue by NE-SmallTown** [babel/babel#12354]
## Bug Report

- [ ] I would like to work on a fix!

**Current behavior**

See [this](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&spec=false&loose=false&code_lz=GYVwdgxgLglg9mABMOcAUBKRBvAvgKH1ElgWVUx1yA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Creact%2Cstage-2&prettier=false&targets=&version=7.12.3&externalPlugins=)

**Input Code**

```js
function foo() {}

function foo() {}
```

Will get `SyntaxError: Identifier 'foo' has already been declared` error

**Expected behavior**

No error

**Babel Configuration (babel.config.js, .babelrc, package.json#babel, cli command, .eslintrc)**

See REPL above

**Environment**

```
REPL
```
- Babel version(s): REPL
- Node/npm version: REPL
- OS: Windows 10
- Monorepo: no
- How you are using Babel:  REPL

**Possible Solution**

**Additional context**

**Comment by JLHwung** [comment]
This is expected. By default Babel assumes `sourceType` is `module` and duplicate function declaration is a syntax error because module body is always in strict mode. You can specify `sourceType` as `script`.

**Comment by NE-SmallTown** [comment]
Thanks for your clarification.

> because module body is always in strict mode. 

I think this is not about strict mode because I can add strict mode to normal script(`application/javascript`) and it won't throw the error. IMO, this is just about the `type`(i.e `module` or `application/javascript`). 

But here the question is, usually we use a `module` as a script(`application/javascript`) after it's been compiled by Babel. So I think maybe we should rely on something like `targetType` rather than `sourceType`?

**Comment by loganfsmyth** [comment]
You are correct, this is not about strict mode, but it is about modules. 

> But here the question is, usually we use a module as a script(application/javascript) after it's been compiled by Babel. So I think maybe we should rely on something like targetType rather than sourceType?

The error here is because the `input` file is parsed as a module, and modules do not allow the code you've provided. If you want to not treat the input as a `module` then you need to set `sourceType`. The way the file is used after it is processed by Babel is up to you once Babel has output the file somewhere.

**Comment by NE-SmallTown** [comment]
Ah, I think this is a interesting case.

If we think finally our code is running on Browser ranther than Babel, then it should not throw error if the Browser doesn't

The key point is that when a developer declare the input file `sourceType` to `module`, does that means or can we confirm he will use it as `type=module` in the browser? Probably not, even from the reality I guess most developers will finally use the bundle or compiled ouput as `type=application/javascript`.

**Comment by nicolo-ribaudo** [comment]
> The key point is that when a developer declare the input file `sourceType` to `module`, does that means or can we confirm he will use it as `type=module` in the browser?

No, this means that the input code should have the same semantics as if it was run as a native module.
Similarly, if you use `class` in your input code many old browsers won't directly run it, but Babel makes sure that it runs with similar semantics.

If you run your input untranspiled code in a browser _as a module_ it throws because of the duplicated declaration: if you tell Babel that your code should be considered a module, it should throw.

**Comment by NE-SmallTown** [comment]
> No, this means that the input code should have the same semantics as if it was run as a native module.

Yes, I agree. But from reality, it's not. 

The current progress is, write a module -> pass it to Babel -> get the result -> use the result as `<script type='application/javascript'>`

So, we don't want to set the input file as `script` because we want to use the `module` syntax(`import/export`) when we are developing an app or library. Therefore, we would use `module`, but finally we will put the result into `<script type='application/javascript'>` and it **won't** throw. If Babel doesn't throw, it's still ok becasue if finally we put the result into `<script type='module'>` and we **will** get the error by Browser. We **won't miss** the error as Babel doesn't throw


> Similarly, if you use class in your input code many old browsers won't directly run it, but Babel makes sure that it runs with similar semantics.

I thinks this is different. Especially for SyntaxError

**Comment by nicolo-ribaudo** [comment]
You are saying that it's possible for Babel to accept invalid code and still produce code that runs.

That's true, but it's out of scope for the project. Our goal is to align to the specification so that when gradually disabling Babel transform over time your code doesn't suddenly start failing.

It's _not_ ok to silently shallow an error, because that error is mandated by the spec.

That said, you can pass the `errorRecovery: true` parser option which doesn't throw for early errors:
https://runkit.com/nicolo-ribaudo/babel-babel-12354

**Comment by NE-SmallTown** [comment]
> Our goal is to align to the specification so that when gradually disabling Babel transform over time your code doesn't suddenly start failing.

That's fair.

By the way, could we add more info when throw the error? For example, change _"SyntaxError: Identifier 'foo' has already been declared"_ to _"SyntaxError: Identifier 'foo' has already been declared, **this is not allowed in a ESM**"_
